### PR TITLE
Detect cyclic fragment references

### LINKED ISSUE #5224: generateServiceApolloSources crashes if a fragment definition references itself
### Version

3.8.2

### Summary

If a fragment description references itself, the code generation step crashes with a stack overflow.

For example, this would crash.
```
fragment CommentFragment on Comment {
  id
  createdAt
  parent {
    ...CommentFragment
  }
}
```

### Steps to reproduce the behavior

_No response_

### Logs

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':cache:generateServiceApolloSources'.
> java.lang.StackOverflowError (no error message)
```
```
Cause: java.lang.StackOverflowError
	at com.apollographql.apollo3.relocated.kotlin.collections.EmptyList.iterator(SourceFile:1)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.GqldirectiveKt.findDeprecationReason(SourceFile:1)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:48)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:227)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:99)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:99)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:227)
	at com.apollographql.apollo3.relocated.com.apollographql.apollo3.ast.internal.ExecutableValidationScope.validate(SourceFile:99)
...
```

### PR DESCRIPTION
**Pull request by martinbonnin**
Closes #5224

### CODE REVIEW

File: libraries/apollo-ast/api/apollo-ast.api [hunk 1]
```diff
@@ -968,17 +1010,38 @@ public final class com/apollographql/apollo3/ast/TransformResult$Replace : com/a
...
 	public fun <init> (Ljava/lang/String;Lcom/apollographql/apollo3/ast/SourceLocation;)V
 }
 
-public final class com/apollographql/apollo3/ast/ValidationDetails : java/lang/Enum {
-	public static final field DuplicateTypeName Lcom/apollographql/apollo3/ast/ValidationDetails;
-	public static final field Other Lcom/apollographql/apollo3/ast/ValidationDetails;
-	public static final field UnknownDirective Lcom/apollographql/apollo3/ast/ValidationDetails;
-	public static fun getEntries ()Lkotlin/enums/EnumEntries;
-	public static fun valueOf (Ljava/lang/String;)Lcom/apollographql/apollo3/ast/ValidationDetails;
-	public static fun values ()[Lcom/apollographql/apollo3/ast/ValidationDetails;
```
Comment (martinbonnin) [inline libraries/apollo-ast/api/apollo-ast.api:981]: There are a bunch of breaking changes in the AST API. I expect them to be mostly harmless. Apollo AST in 3.x was mostly in preview.

File: libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt [hunk 2]
```diff
@@ -28,53 +31,102 @@ import com.apollographql.apollo3.ast.GQLType
...
-   * As the tree is walked, variable references will be put here
+   * These are scoped to the current operation/fragment being validated
    */
   private val variableUsages = mutableListOf<VariableUsage>()
 
-  private val deferDirectiveLabels = mutableMapOf<String, SourceLocation?>()
-  private val deferDirectivePathAndLabels = mutableMapOf<String, SourceLocation?>()
+  private fun GQLFragmentDefinition.detectCycles(globallyVisitedFragments: MutableSet<String>,
+                                                 locallyVisitedFragments: MutableSet<String>,
+                                                 path: List<String>) {
```
Comment (martinbonnin) [inline libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt:78]: This is the actual fix

File: libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt [hunk 3]
```diff
@@ -282,13 +274,10 @@ internal class ExecutableValidationScope(
       return
     }
 
-    fragmentDefinition.selections.validate(fragmentTypeDefinition, this@validate, path)
```
Comment (martinbonnin) [inline libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt:285]: Most of the validations do not recurse in fragment spreads now. Exceptions are 

* `@defer` validations
* fieldsInSetCanMerge

They are handled separately

File: libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/ir/IrOperationsBuilder.kt [hunk 4]
```diff
@@ -316,7 +318,7 @@ internal class IrOperationsBuilder(
   private fun GQLFragmentDefinition.toIr(): IrFragmentDefinition {
     val typeDefinition = schema.typeDefinition(typeCondition.name)
 
-    val inferredVariables = inferVariables(schema, allFragmentDefinitions, fieldsOnDisjointTypesMustMerge)
+    val inferredVariables = inferFragmentVariables(this)
```
Comment (martinbonnin) [inline libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/ir/IrOperationsBuilder.kt:321]: I haven't benchmarked this but this should save a non-trivial amount of work by collecting inferred variables during initial validation.

**Review by martinbonnin — COMMENTED** [review]
Updated this PR to rework a bit the validation:

* Most of the validation do not recurse into fragment spreads
* Moved the decision of what is an error/warning to the callsite
* Added a validation for unused fragments

As a follow up, might be interesting to move the custom apollo-compiler validations to apollo-ast

**Review by BoD — APPROVED** [review]
🚀

File: libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt [hunk 5]
```diff
@@ -580,12 +516,28 @@ internal class ExecutableValidationScope(
...
+    issues.add(
+        DifferentShape(
+          message = buildMessage(fieldA, fieldB,"they have different shapes"),
+          sourceLocation = fieldA.sourceLocation,
+        )
+    )
+    // Also add the symmetrical error
+    issues.add(
+        DifferentShape(
+          message = buildMessage(fieldA, fieldB,"they have different shapes"),
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/ExecutableValidationScope.kt]: Thanks!

**Pull request merged**